doc/ztp/ztp_hld.md

```diff
@@ -0,0 +1,135 @@
+        gh Level Design
+# Requirements
+- Build option to have ZTP enabled
+- SNMP should be available through the ZTP service
+- ZTP should verify, download and install SONiC image
+- Updategraph handles all configuration during ZTP service
+- ZTP should receive post config validation script, that collects switch info and sends it back to a remote location for processing
+- ZTP status may be logged through syslog
+- Interruption of ZTP service should disable the service gracefully
+
+# Flow
+For ZTP using DHCP, provisioning initially takes place over the management network and is initiated through a DHCP hook. A DHCP option is used to specify a configuration script. This script is then requested from the Web server and executed locally on the switch.
+1.	Simplify installation of switch, the steps involved will be
+    - Rack and Stack 
+    - Connect 
+    - Power-on 
+2.	Onie Boots into SONiC (with ZTP enabled) from flash
+    - Power-on ZTP service will take over to load the initial configuration.
+    - ZTP service flag $enabled is true
+    - ZTP service flag $post_install is false
+    - ZTP service flag $post_config is false
+3.	The switch now enters ZTP mode and does the following,
+    - Obtains an IP address from the DHCP server
+    - Obtains the URL for the SONIC image, provisioning script, and post config validation script 
+4.	At this step, the switch will have information to reach the HTTP / TFTP server information by DHCP option 66 or Option 240.
+5.	ZTP service will relay to updategraph service
+    - Within updategraph, ztp_enabled is true, and post_install is false
+    - This triggers updategraph to configure the device with an default config
+    - Upon exiting updategraph, SNMP will start
+6.	ZTP service will now download the software image file.
+7.	ZTP service will validate image to be a SONIC image compatible with the current platform.
+8.	ZTP service will check remaining hard disk for space before doing image install.
+9.	ZTP service will install image using sonic-installer
+    - ZTP service flag $post_install is true
+10.	ZTP service will enable updategraph
+11.	ZTP service will call for reboot to apply new SONIC image, with option ZTP-enabled –post_install
+12.	Upon reboot, if ZTP service flag $post_install is true, ZTP hands off to updategraph
+13.	The configuration script is received and applied by updategraph, returns to ZTP service
```

Comment:
updategraph seems very minigraph like in name, what is the config file being downloaded? a shell script? a minigraph.xml file? a config.json file?

Reply:
Historically updategraph was used to deploy minigraph.xml files but moving forward SONiC configuration will be done using a config.json file.

Reply:
Would it be possible to consider loading a shell script and executing it rather than a config.json? Reason is the following: with FRR the config no longer lives in config.json but in a separate frr.conf file, same with snmp config. Having a bash script allows the operator to have more control over the ZTP process and be creative rather than being forced into a static model. The shell script could replace j2 templates for eg. perform certain daemon reloads etc.
I do understand that with the config checking and reporting that happens later in the process a minimal config can be loaded then an external agent could come in and finish the full device commissioning .

Reply:
Thank you for your input Michel.  The tentative agenda for the next community meeting is on ZTP, we can bring up this point then and see what the community says

Reply:
+1 on what Michel said.